uk.co.compucorp.civicrm.tasksassignments/CRM/Tasksassignments/BAO/Document.php

```diff
@@ -1,35 +1,136 @@
-            }
-            if (empty($params['status_id'])) {
-                $params['status_id'] = 1;
-            }
-        }
-
-        CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
-
-        $instance = parent::create($params);
-        CRM_Tasksassignments_Reminder::sendReminder((int)$instance->id);
-        
-        CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
-        
-        return $instance;
+class CRM_Tasksassignments_BAO_Document extends CRM_Tasksassignments_DAO_Document {
+
+  /**
+   * Create a new Document based on array-data
+   *
+   * @param array $params key-value pairs
+   * @return CRM_Tasksassignments_DAO_Document|NULL
+   */
+  public static function create(&$params) {
+    $entityName = 'Document';
+    $hook = empty($params['id']) ? 'create' : 'edit';
+
+    if ($hook === 'create') {
+      // If creating a new Document, we require all three contacts to be defined.
+      if (empty($params['source_contact_id']) || empty($params['target_contact_id']) || empty($params['assignee_contact_id'])) {
+        throw new CRM_Exception("Please specify 'source_contact_id', 'target_contact_id' and 'assignee_contact_id'.");
+      }
+      if (empty($params['status_id'])) {
+        $params['status_id'] = 1;
+      }
+    }
+
+    CRM_Utils_Hook::pre($hook, $entityName, CRM_Utils_Array::value('id', $params), $params);
+
+    $instance = parent::create($params);
+    CRM_Tasksassignments_Reminder::sendReminder((int) $instance->id);
+
+    CRM_Utils_Hook::post($hook, $entityName, $instance->id, $instance);
+
+    return $instance;
+  }
+
+  /**
+   * Clone documents if there are 'days_to_create_a_document_clone' days
+   * before their expire date.
+   * Return a count of cloned documents or NULL if 'days_to_create_a_document_clone'
+   * is zero.
+   * 
+   * @return int|NULL
+   */
+  public static function cloneDocuments() {
+    $daysToCreateADocumentClone = self::getDaysToCreateADocumentClone();
+    if (!$daysToCreateADocumentClone) {
+      return NULL;
+    }
+    $date = (new DateTime())
+            ->add(new DateInterval('P' . $daysToCreateADocumentClone . 'D'))
+            ->format('Y-m-d');
+    return self::cloneApprovedDocumentsWithExpireDate($date);
+  }
+
+  /**
+   * Return a 'days_to_create_a_document_clone' value from T&A Settings.
+   * 
+   * @return int
+   */
+  protected static function getDaysToCreateADocumentClone() {
+    $setting = civicrm_api3('TASettings', 'getsingle', array(
+      'sequential' => 1,
+      'fields' => "days_to_create_a_document_clone",
+    ));
+    return !empty($setting['value']) ? $setting['value'] : 0;
+  }
+
+  /**
+   * Search for approved Documents with given expire date and create their clone.
+   * Return a count of cloned documents.
+   * 
+   * @param string $expireDate (date in Y-m-d format)
```

Comment:
Recently, I've been thinking that, in cases like this, it's better if we enforce the use of DateTime objects instead of expecting a string with a date in the right format. Just like I did here: https://github.com/civicrm/civihr/pull/1104/commits/a0dd0dfc716d50a7f768c39d03f48475310c321c#diff-e17c168c0f4c4c52b05295b72cca13f6R281

These are the advantages I see on this approach:
1. The format gets irrelevant, which is good! Imagine that, for some reason, you need to change the format to Ymd instead of Y-m-d. Without a DateTime, you would need to change the dates on every call to this method
2. Implicit validation. You don't need to do date validation on your code to check if you received a valid date (Speaking of that, you didn't do it for this method)

Also, on cloneDocuments() (the method that calls cloneApprovedDocumentsWithExpireDate()) you already have a DateTime object